File: apps/mobile/shared/Sources/FlagshipCore/PairingOrderEnvelope.swift
import Foundation

/// Swift mirror of `packages/protocol/src/pairingOrder.ts` `pairingOrderToJson` —
/// the PLAINTEXT `add-paired-session` envelope JSON for the SECRET-FREE pairing
/// path (the twin of `SwkDelivery`, for the paired-session token).
///
/// The first recipe carries ZERO pairing secrets. The owner-IRK-signed
/// `add-paired-session` order is serialized into this plaintext `{request,
/// signature}` JSON, which the box parses verbatim. It is either:
///   - EMBEDDED in the recipe as the unsigned `pairingOrder` sibling (offline),
///     or
///   - SEALED to the box's directory identity + deposited on `.com`'s blind
///     pairing-deposit lane post-registration (default online).
///
/// The box re-derives canonical bytes from the field VALUES, so key order is
/// irrelevant to verification — but we reproduce the pinned cross-platform
/// vector's key order
/// (`{request:{type,serverId,token,issuedAt}, signature}`) byte-for-byte
/// (`packages/protocol/tests/pairingOrder.test.ts`), so the embedded sibling +
/// the sealed deposit payload are identical to the TS / webapp / Kotlin twins.
public enum PairingOrderEnvelope {
    /// Serialize an `add-paired-session` order + its owner-IRK signature into the
    /// plaintext envelope JSON string (UTF-8). Built by hand (NOT
    /// `JSONSerialization`, which doesn't preserve key order) to be byte-identical
    /// to the TS `pairingOrderToJson` pinned vector.
    public static func toJson(order: AddPairedSessionOrder, signatureHex: String) -> String {
        // JSON.stringify emits no whitespace, strings double-quoted with the
        // standard JS escapes, and `issuedAt` as a bare number. The order's
        // string fields are field-guarded at sign time (no '|'/control chars);
        // serverId/token are hostname/hex.
        let req = "{\"type\":\"add-paired-session\","
            + "\"serverId\":\(jsonString(order.serverId)),"
            + "\"token\":\(jsonString(order.token)),"
            + "\"issuedAt\":\(order.issuedAt)}"
        return "{\"request\":\(req),\"signature\":\(jsonString(signatureHex))}"
    }

    /// Minimal JSON string encoder matching `JSON.stringify` for the characters
    /// that can appear in these fields (`"`, `\`, and control chars). Mirrors
    /// the ECMAScript string-serialization rules.
    static func jsonString(_ s: String) -> String {
        var out = "\""
        for scalar in s.unicodeScalars {
            switch scalar {
            case "\"": out += "\\\""
            case "\\": out += "\\\\"
            case "\u{08}": out += "\\b"
            case "\u{0C}": out += "\\f"
            case "\n": out += "\\n"
            case "\r": out += "\\r"
            case "\t": out += "\\t"
            default:
                if scalar.value < 0x20 {
                    out += String(format: "\\u%04x", scalar.value)
                } else {
                    out.unicodeScalars.append(scalar)
                }
            }
        }
        out += "\""
        return out
    }
}

File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/core/PairingOrderEnvelope.kt
package com.flagshipserver.app.core

/**
 * Kotlin mirror of `packages/protocol/src/pairingOrder.ts` `pairingOrderToJson` —
 * the PLAINTEXT `add-paired-session` envelope JSON for the SECRET-FREE pairing
 * path (the twin of [SwkDelivery], for the paired-session token).
 *
 * The first recipe carries ZERO pairing secrets. The owner-IRK-signed
 * `add-paired-session` order is serialized into this plaintext `{request,
 * signature}` JSON, which the box parses verbatim. It is either EMBEDDED in the
 * recipe as the unsigned `pairingOrder` sibling (offline) or SEALED to the box's
 * directory identity + deposited on `.com`'s blind pairing-deposit lane
 * post-registration (default online).
 *
 * Built by hand (NOT kotlinx.serialization, whose escaping/ordering we don't want
 * to depend on) to be byte-identical to the TS `pairingOrderToJson` pinned vector
 * (`packages/protocol/tests/pairingOrder.test.ts`): key order
 * `{request:{type,serverId,token,issuedAt}, signature}`, no whitespace,
 * `issuedAt` a bare number.
 */
object PairingOrderEnvelope {
    fun toJson(
        serverId: String,
        token: String,
        issuedAt: Long,
        signatureHex: String,
    ): String {
        val req = "{\"type\":\"add-paired-session\"," +
            "\"serverId\":${jsonString(serverId)}," +
            "\"token\":${jsonString(token)}," +
            "\"issuedAt\":$issuedAt}"
        return "{\"request\":$req,\"signature\":${jsonString(signatureHex)}}"
    }

    /** Minimal JSON string encoder matching `JSON.stringify` for the characters
     *  that can appear in these fields (`"`, `\`, and control chars). */
    fun jsonString(s: String): String {
        val out = StringBuilder("\"")
        for (c in s) {
            when (c) {
                '"' -> out.append("\\\"")
                '\\' -> out.append("\\\\")
                '\b' -> out.append("\\b")
                '\u000C' -> out.append("\\f")
                '\n' -> out.append("\\n")
                '\r' -> out.append("\\r")
                '\t' -> out.append("\\t")
                else -> if (c.code < 0x20) out.append("\\u%04x".format(c.code)) else out.append(c)
            }
        }
        out.append("\"")
        return out.toString()
    }
}
